tests/conftest.py

```diff
@@ -205,7 +205,19 @@ def pytest_configure():
         from pybind11_tests.eigen import have_eigen
     except ImportError:
         have_eigen = False
-    pypy = platform.python_implementation() == "PyPy"
+
+    # Provide simple `six`-like aliases.
+    pytest.PY2 = (sys.version_info.major == 2)
+    pytest.PY3 = (sys.version_info.major == 3)
+    pytest.CPYTHON = (platform.python_implementation() == "CPython")
+    pytest.PYPY = (platform.python_implementation() == "PyPy")
```

Comment:
Should we really be putting stuff in someone else's module?

Reply:
Same here. I agree with @bstaletic, that this should probably not be part of `pytest`. `PY2`, etc, in full-caps, should be clear enough that no one uses the same name? It's already better than using lower-case `pypy`, anyway.

Reply:
> it at least sticks with convention.

Oh, is this a convention?

Reply:
I think @EricCousineau-TRI means the six convention. I don't think injecting things into pytest is convention; and if it is, it's a bad convention. One reason conftest runs before but in the same namespace as tests is to allow you _not_ to have to communicate through pytest. If you want to replicate the six convention, I would do this:

```python
class six:
    PY2 = sys.version_info.major == 2
```
I would actually _not_ add PY3, as that would force the correct usage of PY2 - the things currently checking against `PY3` should actually be checking `not PY2`. Python 4 won't change the things we are checking PY3 for.

Reply:
If we're going for `six` as a name anyway, why not just add it to `requirements.txt` and depend on it?

Reply:
> I would actually _not_ add PY3, as that would force the correct usage of PY2 - the things currently checking against `PY3` should actually be checking `not PY2`. Python 4 won't change the things we are checking PY3 for.

Good point as well, btw! :+1:

Reply:
We could do that, but unlike a normal package, we don't have an "install" to do the testing, so this would mean everyone running the tests would have to install six. It's [pretty common](https://github.com/tomerfiliba/plumbum/blob/6788d15b017d7d6a3cf1834c9b098db2719762a9/plumbum/lib.py#L34-L94) to implement just the bit of six you need - and for pybind11, this seems like the simplest way to do this (since we really only require a single dependency, pytest, currently to run the tests).

Reply:
A agree, making `six` a requirement causes more harm (lost time) than good, especially this late in the game.

To @YannickJadoul and @bstaletic 's concerns: I feel sticking with the precedent (`pytest.pypy`) is the right thing here. Going through the trouble establishing a separate module for this is again more trouble than it is worth it IMO. I think it's important to be able to make meaningful small steps in the right direction, without tearing open a whole new construction site every time.

I really like @YannickJadoul 's suggestion to only introduce `PY2` but NOT `PY3`. Hopefully we'll never have to go through something like the 2-to-3 transition again.

@YannickJadoul : if `pytest.PY2` is centrally available I definitely prefer that over my `str is bytes` test. (My thinking was: if there is nothing central, I use a test that it most directly related to what I'm dealing with, but I think the central variable is best.)

Reply:
Great, let's get this in, then!

I'm still not a huge fan of adding to pybind11 (I do think adding a separate module and importing it could be cleaner; and it should be painless, just a few constants in a separate file), but it does work, so it's not a crucial point for me.

> I really like @YannickJadoul 's suggestion to only introduce `PY2` but NOT `PY3`. Hopefully we'll never have to go through something like the 2-to-3 transition again.

@henryiii's idea, btw ;-)

Reply:
It would be more appropriate to use a [pytest fixture](https://docs.pytest.org/en/stable/fixture.html) for this instead of modifying the namespace?

```python
# conftest.py

@pytest.fixture
def compat():
  return SomeObject()

```

```python

def test_a_thing(compat):
   if compat.PY2:
      # do something
```

You could also add boolean PY2/etc fixtures:

```
@pytest.fixture
def PY2():
   return True if whatever else False

def test_a_thing(PY2):
   if PY2:
      blah

Reply:
Fully agree with @virtuald (though markers on tests would be even better in many cases). PyTest uses fixtures and markers, not injection into the pytest namespace. If "pytest.PY3" was ever added, this might break.

Reply:
Yes, but to be honest, custom markers/fixtures seem to be too complex/elaborate for what this is (currently) doing, no? If the alternatives are between `pytest` and custom markers, I think I'd (currently) prefer the first one.
But even better, I'd just go for some auxiliary module and `from auxiliary_module import *`? I don't think there's any reason to use the `pytest_configure` hook for this, is there?

Reply:
Nah, I think the injection into `pytest` is weird, but my vote is we keep things relatively simple in this PR, then go the auxiliary module route (`from pybind_test_helpers` or whatevs) in the future.